### DISCUSSION TITLE
Refactor Mapping yaml

### DISCUSSION THREAD

**Discussion by jwildfire** [Gilead-Public/gsm#1771] [Ideas]
# Overview

Consider refactoring the `inst/data_mapping.yaml` workflow to incorporate the new `spec` property and more easily support data-requirements from multiple modules/reports.

# Current Implementation

Currently `data_mapping.yaml` uses `RunQuery()` to execute a series of simple sql commands. including: 
- "SELECT subjectname as subject_nsv, * FROM df WHERE data_entry_lag > 10"
-  "SELECT subjectenrollmentnumber as subjid, * FROM df WHERE deemedimportant == 'Yes'"

The current implementation includes the following: 
1. Selecting columns
2. Renaming columns
3. Filtering based on a single value
4. Filtering based on multiple values
5. Filtering based on a numeric comparison
6. Filtering on multiple conditions

# Possible Refactor

Provide a machine-readable specification (YAML/List) that would easily allow requirements from multiple modules to be combined in to a single mapping step. We'd need to support all 6 use cases above (and possibly others).  Select and Rename (1. and 2. above) can probably be accomplished using the current `spec` objects. We'd just need to add an optional `rename` parameter for each required column. 

Adding filters is also pretty straightforward, I think. For example, the current implementation of `Mapped_Enrolled`: 

```
spec:
  Raw_SUBJ:
      subjectid:
        required: true
      enrollyn:
        required: true
steps: 
  - name: RunQuery
    output: Mapped_Enrolled
    params:
      df: Raw_SUBJ
      strQuery: "SELECT subjectid as raw_subjectid, * FROM df WHERE enrollyn == 'Y'"
```

becomes: 

```
spec:
  Raw_SUBJ:
      _filters:
        - "enrollyn == 'Y'"
      subjectid:
        required: true
        rename: raw_subjectid
      enrollyn:
        required: true
steps: 
  - name: RunQuery
    output: Mapped_Enrolled
    params:
      df: Raw_SUBJ
     spec: spec
```

`RunQuery` then gets a refactor to parse `{spec$df}` into a query. We can discuss whether to stick with SQL or change to DuckDB or just use dplyr. 

In fact, we can just create `RunQueries()` that loops through `spec` and calls `RunQuery` for each element: 

```
steps:
   - name: RunQueries
   - output: lMapped
   - params: 
      - lData: raw_data
      - lSpec: spec
```

**Comment by jwildfire** [comment]
@samussiah @dpastoor @lauramaxwell @jonthegeek Let me know if you have any initial thoughts. Going to work on a prototype.

**Comment by jwildfire** [comment]
Implementation above isn't quite right since `_filter` and `rename` really apply to output data (e.g. `Mapped_SAE`), but `spec` is associated with input data (e.g. Raw_AE). This causes issues when there multiple mapped data sets generated from a single  raw data set. For `example _filter: "aeser == 'Y'"` applies to `Mapped_SAE` but not to `Mapped_AE`, so doesn't really make sesnse to have it nested under `spec$Raw_AE$_filters`

Basically we need both input specs and output specs. I'll think about the best way to address.

**Comment by dpastoor** [comment]
One thing that I like that this is more directly taking on is expressing what is needed within the data to allow us to be much more explicit around fetching data as needed rather than the current `*` usage.

When considering the query component, the sticks out to me when looking at this is there is a good bit of "prior art" around building these types of sql dsl's and mappings, I worry that this may trend into the territory of building a little mini-ecosystem (and the need to maintain) in an area that we could take advantage of existing solutions off the shelf. 

That said, even if we do go down the track of surgically designing a tightly scoped dsl to maintain, it would still be worth looking at such implementation such as https://docs.getdbt.com/reference/model-properties. 

For what its worth, within the current DSL, though its "simple" by being compacted into a single sql strings, its also amenable to translating to other implementation language(s) - but if as we start coming up with such dialects, its going to couple the implementation and execution of these pipelines much more tightly to R, if just purely the energy to write the configuration adjustments and testing everything out. There are also many examples of "compilers", verifiers and linters that help generate models out of sql, https://sqlc.dev/  https://docs.sqlc.dev/en/latest/howto/verify.html and linter: https://github.com/sqlfluff/sqlfluff

It feels like we really need to make a decision sooner rather than later around being based on SQL, and taking advantage of tools in that ecosystem for much of the data querying/mapping _or_ own that likely we are going to do enough smoothing out and adjustments in R and lean into R. Being in the middle risks getting the worst challenges of both worlds that we have to paper over.

Last immediate statement around separating out sql style filters into arrays or the like that can become challenging is if there ever start to be cases where you need to differentiate between _and_ clauses vs _or_ clauses, and is where some of the yaml dsls start to get cumbersome.

I'll likewise keep thinking on it, and one area that would be very helpful is to get a better sense of what these column mappings/remappings we'd likely be expecting to see, and a sense of how frequent these would be. Eg will we pretty consistenly be renaming columns compared to what comes out of rawplus/etc or will much of that be consistent and the column renames will be an occasional activity.

**Reply by jwildfire** [reply]
> It feels like we really need to make a decision sooner rather than later around being based on SQL, and taking advantage of tools in that ecosystem for much of the data querying/mapping or own that likely we are going to do enough smoothing out and adjustments in R and lean into R. Being in the middle risks getting the worst challenges of both worlds that we have to paper over.

Agree with this for sure, and do feel like I'm a little bit stuck in a middle ground. If we decide to go with R we could look more closely at some of the tools I listed in #1744. I think several of them could work for us, but do come with some overhead. 

I don't know the full sql ecosystem (or arrow/duckdb) very well so will look into your links a bit and think about if they would work for our use case. Also thinking it might be helpful to put together a slightly detailed deck/discussion describing our key use cases.

**Comment by jwildfire** [comment]
@dpastoor I think the main issue I'm wrestling with is how to combine specifications/queries from different reporting modules. I'm going to try to write up a simple example below focused on AEs. 

# Overview
Let's say we have a safety study that is requesting 3 central monitoring outputs related to safety:
1. KRI Report with 2 metrics using the gsm analytics framework
2. KRI Deep Dive app with site- and participant-level drilldown showing key AE details
3. A Custom AE Report (e.g. [aeexplorer](https://github.com/RhoInc/aeexplorer))

The data flow for these reports would be something like this: 

![image](https://github.com/user-attachments/assets/b3cf759d-8d72-4c4e-b66b-e90b23528f72)

# Module Data Requirements

So, what are the data requirements for the AE domain in these 3 reports? Let's just list the sql calls needed to generate mapped data. 
1. The KRI report just needs to trace back the requirements for generating each of the KRIs. To generate the KRI, we really just need columns one record per AE with columns for for Subject ID and Seriousness. This is well implemented in {gsm} the sql calls are easy enough: 
   - dfAE: "SELECT subjid FROM df"
   - dfSAE: "SELECT subjid FROM df WHERE aeser = 'Y'"
2. The KRI deep dive requires additional data because we want to be able to see AEs for given sites/participants (hence the dotted line from the Deep Dive app to the Mapped Data". Let's keep it simple and just add a few things like start date and end date for now:
   - dfAE: "SELECT subjid, aeser, aestdate, aeenddate  FROM df"
   - dfSAE: "SELECT subjid, aeser, aestdate, aeenddate FROM df WHERE aeser = 'Y'"
3. As shown above, the AE explorer doesn't use the reporting or analysis framework at all, and it doesn't need an SAE domain. Data requirements are shown [here](https://github.com/RhoInc/aeexplorer/wiki/Data-Guidelines#default-chart) - at a minimum it needs subject id, system organ class and preferred term. Our sql query might be: 
   - dfAE  "SELECT subjid, aesoc, aept  FROM df"

# Combined Requirements

Combining those requirements is easy enough for this simple use case. At a minimum, we need: 
- dfAE: "SELECT subjid, aeser, aestdate, aeenddate, aesoc, aept  FROM df"
- dfSAE: "SELECT subjid, aeser, aestdate, aeenddate, aesoc, aept  FROM df WHERE aeser = 'Y'""

So the big question is how do we document the requirements for each module in a way that they can easily be combined? Is there an existing framework for this? Or maybe it's not even needed in DuckDB world? 

## Notes
- I've left out the requirements around renaming columns and having optional columns for now. I think those should be relatively straightforward to deal with once we align on an overall framework for filter/select. 
- dfSAE is entirely contained in dfAE for all of these examples, so maybe we just filter when we generate Analysis$dfInput? Or maybe it really doesn't matter in DuckDB land? 
- It wouldn't be too hard to add optional tidyr filter expressions `strNumeratorFilter` and `strDenominatorFilter` to `input_rate()` and just abstract this nuance out of the data model entirely.

**Reply by jwildfire** [reply]
Another lovely answer would be, "We don't need to combine the specs because, technology X just deals with this for us. We can just write sql transforms for each module and {gsm.template} will do the rest" 

That might leave some inefficiencies in an R-only implementation (maybe mapping has to be re-run for each report/app), but that might be ok as long as we explain how a more efficient implementation works.

**Comment by jwildfire** [comment]
The more I think about it, the more I think I'm over-engineering this process. I'm going to try to summarize how I see the study configuration process working, and then summarize implementation. 

# Study Configuration of Data Mapping

I think the data mapping for each new study using {gsm} could be configured roughly as follows: 

1. The study team should select the reports and metrics that are needed for the study. 
2. Each of those reports and metrics should have well documented data requirements (likely a yaml) that specifies which tables and columns are required and optional. Basically like the `spec` object above. 
    - Could also consider adding more complex "rules" to these specs using one of the R packages listed in #1744, but not in v2.1. 
4. Those specs should be combined in to a "study specification" 
5. That study specification should be converted to a script (e.g. a `mapping_data.yaml` workflow) that pulls/transforms all required raw data into a mapped database layer. **This script can just be generated manually for each study for now.** 
    - Parts of this can probably be automated, but I'm leaning towards leaving this out of scope for v2.1. We can always add utility functions down the road. 
6. Whenever the mapping script is run, the output can be checked against the study specification to make sure all required columns are provided (and in the correct format). 

# Implementation Plan

To implement this, we need to:
- We need to think about how (1) actually happens. Would be nice to have a master list of all OpenRBQM reports/metrics somewhere. Maybe that goes in {gsm.template}. Maybe we do want to keep {gsm.library}. Maybe we can just make a website over as part of {OpenRBQM}. Might be worth a separate discussion diving in to this ... 
- We need to create the specs described in (2). 
   - This is basically done for the metrics. 
   - I think we probably need to create an `inst/workflow/reports` folder that has a separate yaml with a spec for each report. 
      - This will require updates in `inst/workflow/reports.yaml` and `snapshot.yaml`.
   - We'll likely want to add this `inst/workflow/reports` folder in our extension modules as well, starting with `gsm.app` and `gsm.endpoints`
- (3) is basically done using `CombineSpecs()`. Issues are already in to update. 
- re: (4) We should probably update `inst/workflow/mapping_data.yaml` to align with the basic spec. 
   - we should think about whether to include data needed to support for {gsm.app} and {gsm.endpoints} here in {gsm}. 
   - Maybe there's an argument that we should move the workflows to the proposed {gsm.template} repo
   - Maybe we also move some of the `util` functions related to workflow to {gsm.template} ... or maybe not. Worth a quick discussion though. 
- Basic implementation of (5) is in using `CheckSpec()`. 

Last but not least, all of this needs to be written up in a vignette. 

# tl;dr

We need to: 
- Define basic specs for metrics and reports as part of their workflows. 
- Combine those specs into a study-level mapping specification
- Write a mapping script to generate the mapping data
- Check output of the mapping script against the mapping spec.
